cli/src/main/java/com/box/l10n/mojito/cli/command/checks/GlossaryTermTrie.java

```diff
@@ -0,0 +1,94 @@
+package com.box.l10n.mojito.cli.command.checks;
+
+import org.apache.commons.lang3.StringUtils;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.stream.Collectors;
+
+public class GlossaryTermTrie {
+
+    private class Node {
+        Map<String, Node> children = new HashMap<>();
+        List<GlossaryTerm> glossaryTerms = new ArrayList<>();
+    }
+
+    private Node rootNode = new Node();
+
+    public GlossaryTermTrie(List<GlossaryTerm> terms) {
+        for(GlossaryTerm term: terms) {
+            this.add(term);
+        }
+    }
+
+    public void add(GlossaryTerm term) {
+        String[] words = term.getTerm().split("\\W+");
+        Node currentNode = rootNode;
+        for (String word : words) {
+            word = word.toLowerCase();
+            if(!currentNode.children.containsKey(word)) {
+                currentNode.children.put(word, new Node());
+            }
+            currentNode = currentNode.children.get(word);
+        }
+        currentNode.glossaryTerms.add(term);
+    }
+
+    public GlossarySearchResult runGlossaryCheck(String source) {
+        String[] words = source.split("\\W+");
+        GlossarySearchResult result = new GlossarySearchResult(source);
+        List<String> failures = new ArrayList<>();
+        for(int i = 0; i < words.length; i++) {
+            Node current = rootNode;
+            for (int j = i; j < words.length; j++) {
+                if (!current.children.containsKey(words[j].toLowerCase())){
+                    break;
+                }
+                current = current.children.get(words[j].toLowerCase());
+                if(!current.glossaryTerms.isEmpty()){
+                    checkGlossaryTerms(source, result, failures, current);
+                }
+            }
+        }
+        result.failures = failures;
+        return result;
+    }
+
+    private void checkGlossaryTerms(String source, GlossarySearchResult result, List<String> failures, Node current) {
+        String trimmedWhitespace = source.trim().replaceAll(" +", " ");
+        boolean allFailed = true;
+        for (GlossaryTerm glossaryTerm : current.glossaryTerms) {
```

Comment:
i'm not sure to understand in which case will have multiple glossary term for 1 node and, i'm not sure to understand why we need to search again in the string for the whole match since we've done the search using the trie before.

Is this logic to check that case in the source is the same as declared in the glossary term but we lost the info in the try?

Reply:
It's to handle the use case where the glossary contains terms in both cases e.g."Tag" & "tag", in this example both are acceptable so we maintain a list of glossary terms on the node and then check against them. 

The trie stores it's entries in lower case, we then traverse the trie by setting our search words to lower case too, this should guarantee we get a case insensitive match. Once we have the match we then want to check that the term also matches the case, although looking at the code again I don't think I need to find the index of the term in the source string I can just check if the source string contains the glossary term, if not then we know it doesn't match the glossary

Reply:
I was sensing that but then I was a little bit puzzle with the use case then. Since we basically want to enforce a certain case (upper case Tag) then having both is sort of pointless, no? That'd allow tag and Tag and refuse TAg but is that not relevant to check, I mean we could do that for any string then. 

Wondering if that should just be called a "CaseChecker", we're not really checking much from the glossary point of view. I guess more classic glossary check is for example to make sure that the translation uses the translation specified in the glossary but for source only analysis, here we just check casing, right?

Let's discuss more offline.